### ISSUE TITLE
Rustdoc does not format ASCII art containing `###` at the start of a line correctly

### ISSUE THREAD

**Issue by rjvdw** [rust-lang/rust#105527]
I tried this code:

```rust
//! This is some documentation with ASCII art:
//!
//! ```text
//! #..#.####.#....#.....##..
//! #..#.#....#....#....#..#.
//! ####.###..#....#....#..#.
//! #..#.#....#....#....#..#.
//! #..#.#....#....#....#..#.
//! #..#.####.####.####..##..
//! ```
```

I expected to see this happen: When I run `rustdoc`, I expect to see ASCII art in the output that matches the above example.

Instead, this happened: In the output, this actually appears:
```text
#..#.####.#....#.....##..
#..#.#....#....#....#..#.
###.###..#....#....#..#.
#..#.#....#....#....#..#.
#..#.#....#....#....#..#.
#..#.####.####.####..##..
```
It appears that if a line starts with at least three `#`s, one gets "swallowed" somehow.

### Meta

`rustc --version --verbose`:
```
rustc 1.65.0 (897e37553 2022-11-02)
binary: rustc
commit-hash: 897e37553bba8b42751c67658967889d11ecd120
commit-date: 2022-11-02
host: x86_64-apple-darwin
release: 1.65.0
LLVM version: 15.0.0
```

<details><summary>Backtrace</summary>
<p>

```
n/a
```

</p>
</details>

**Comment by camelid** [comment]
This is actually a regression. I haven't bisected, but it occurred sometime between 1.50.0 and 1.63.0 based on a quick test. In 1.50.0, the following code block

`````rust
/// ```markdown
/// # Header
/// ```
``````

rendered as `# Header`, but in 1.63.0 it renders as a blank code block.

**Comment by GuillaumeGomez** [comment]
Thanks for the information @camelid !

**Comment by albertlarsan68** [comment]
The regression is in nightly-2021-07-15

Commits in range:

  commit[0] 2021-07-13UTC: Auto merge of #86827 - camsteffen:hash-lint-resolved, r=oli-obk
  commit[1] 2021-07-13UTC: Auto merge of #87044 - cjgillot:expnhash, r=petrochenkov
  commit[2] 2021-07-14UTC: Auto merge of #87106 - Mark-Simulacrum:edition-no-clone, r=petrochenkov
  commit[3] 2021-07-14UTC: Auto merge of #86211 - tlyu:option-result-overviews, r=joshtriplett
  commit[4] 2021-07-14UTC: Auto merge of #87118 - JohnTitor:rollup-8ltidsq, r=JohnTitor
  commit[5] 2021-07-14UTC: Auto merge of #87082 - michaelwoerister:const-in-debuginfo-type-names-fix, r=oli-obk,wesleywiser
  commit[6] 2021-07-14UTC: Auto merge of #87133 - GuillaumeGomez:rollup-pfz9jbk, r=GuillaumeGomez
  commit[7] 2021-07-14UTC: Auto merge of #86765 - cuviper:fuse-less-specialized, r=joshtriplett


Maybe this happened in #87056 (merged by #87133) (cc @GuillaumeGomez)
